.circleci/config.yml

```diff
@@ -1,75 +1,217 @@
+      - attach_workspace:
+          at: docker-cache
+      - run:
+          name: Discover Docker Tags
+          command: |
+            mkdir -p docker-cache
+            .circleci/bin/docker-tags "$CIRCLE_SHA1" "$CIRCLE_BRANCH" | sed 's/\//-/g' \
+              > docker-cache/docker-tags.txt
       - run:
-          command: .circleci/build.sh
+          name: Docker build
+          command: |
+            docker build \
+              --build-arg TOOL_NODE_FLAGS="--max-old-space-size=4096" \
+              -t "$DOCKER_REPOSITORY:$CIRCLE_SHA1" .
+            mkdir -p docker-cache
+            docker save \
+              -o docker-cache/docker-image.tar \
+              "$DOCKER_REPOSITORY:$CIRCLE_SHA1"
           no_output_timeout: 30m
+      - persist_to_workspace:
+          root: docker-cache
+          paths:
+            - docker-image.tar
+            - docker-tags.txt
 
+  docker-push:
+    <<: *defaults
+    steps:
+      - setup_remote_docker
+      - attach_workspace:
+          at: docker-cache
       - run:
-          command: .reaction/jsdoc/build.sh
-          no_output_timeout: 2m
-
-      # deploy the docs if on master branch
-      - deploy:
-          name: Deploy to S3 if tests pass and branch is Master
+          name: Load Docker Image
           command: |
-            if [ "${CIRCLE_BRANCH}" == "master" ]; then
-              if [[ "${API_DOC_BUCKET}" && "${API_DOC_BUCKET_REGION}" ]]; then
-                aws s3 sync /tmp/reaction-docs ${API_DOC_BUCKET} --delete --region ${API_DOC_BUCKET_REGION}
-              else
-                echo "S3 bucket configuration not found for jsdocs"
-                echo "Set API_DOC_BUCKET and API_DOC_BUCKET_REGION to build and deploy jsdocs to S3"
-              fi
+            docker load < docker-cache/docker-image.tar
+      - run:
+          name: Tag Docker Image
+          command: |
+            cat docker-cache/docker-tags.txt \
+              | xargs -t -I % \
+                docker tag \
+                  "$DOCKER_REPOSITORY:$CIRCLE_SHA1" \
+                  "$DOCKER_REPOSITORY:%"
+      - run:
+          name: Docker Push
+          command: |
+            docker login -u "$DOCKER_USER" -p "$DOCKER_PASS"
+            docker push "$DOCKER_REPOSITORY:$CIRCLE_SHA1"
+            cat docker-cache/docker-tags.txt \
+              | xargs -t -I % \
+                docker push "$DOCKER_REPOSITORY:%"
+  deploy-docs:
+    <<: *defaults
+    steps:
+      - checkout
+      - run:
+          name: Install AWS CLI
+          command: sudo apt-get -y -qq install awscli
+      - run:
+          name: NPM Install JSDoc
+          command: sudo npm install -g jsdoc
+      - run:
+          name: Build JSDoc files
+          command: |
+            jsdoc . \
+              --verbose \
+              --configure .reaction/jsdoc/jsdoc.json \
+              --readme .reaction/jsdoc/templates/static/README.md
+      - run:
+          name: Deploy Doc files to S3 if on Master
```

Comment:
A subtle, non-blocking nitpick... The leftover "if on Master" here indicates (to me) a check within in the step. There's no mention of master in the command as that logic has been moved to the CircleCI workflow. Could this confuse someone on first look?